File: functions/src/gig_search_index.ts
export const GIG_SEARCH_SCHEMA_VERSION = 1;
const MAX_GIG_SEARCH_CHARACTERS = 2500;

function firstNonEmptyString(values: unknown[]): string {
  for (const value of values) {
    if (typeof value === "string" && value.trim().length > 0) {
      return value.trim();
    }
  }
  return "";
}

function asRecord(value: unknown): Record<string, unknown> {
  return value !== null &&
    typeof value === "object" &&
    !Array.isArray(value) ? value as Record<string, unknown> : {};
}

export function normalizeGigSearchText(value: string): string {
  return value
    .trim()
    .toLowerCase()
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/[^a-z0-9]+/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

export function buildGigSearchGrams(
  data: Record<string, unknown>
): string[] {
  const location = asRecord(data.location);
  const source = normalizeGigSearchText([
    firstNonEmptyString([data.title]),
    firstNonEmptyString([data.description]),
    firstNonEmptyString([location.label]),
  ].join(" ")).slice(0, MAX_GIG_SEARCH_CHARACTERS);
  const grams = new Set<string>();
  for (let size = 1; size <= 3; size += 1) {
    if (source.length < size) break;
    for (let index = 0; index <= source.length - size; index += 1) {
      grams.add(source.slice(index, index + size));
    }
  }
  return [...grams].sort();
}

File: lib/src/features/gigs/data/gig_search_index.dart
const int gigSearchSchemaVersion = 1;
const int _maxIndexedSearchCharacters = 2500;

String normalizeGigSearchText(String value) {
  const replacements = <String, String>{
    'á': 'a',
    'à': 'a',
    'â': 'a',
    'ã': 'a',
    'ä': 'a',
    'é': 'e',
    'è': 'e',
    'ê': 'e',
    'ë': 'e',
    'í': 'i',
    'ì': 'i',
    'î': 'i',
    'ï': 'i',
    'ó': 'o',
    'ò': 'o',
    'ô': 'o',
    'õ': 'o',
    'ö': 'o',
    'ú': 'u',
    'ù': 'u',
    'û': 'u',
    'ü': 'u',
    'ç': 'c',
  };

  final lower = value.trim().toLowerCase();
  final folded = lower.split('').map((character) {
    return replacements[character] ?? character;
  }).join();
  return folded
      .replaceAll(RegExp('[^a-z0-9]+'), ' ')
      .replaceAll(RegExp(r'\s+'), ' ')
      .trim();
}

String buildGigSearchText({
  required String title,
  required String description,
  Map<String, dynamic>? location,
}) {
  final normalized = normalizeGigSearchText(
    [title, description, location?['label']?.toString() ?? ''].join(' '),
  );
  if (normalized.length <= _maxIndexedSearchCharacters) return normalized;
  return normalized.substring(0, _maxIndexedSearchCharacters);
}

List<String> buildGigSearchGrams({
  required String title,
  required String description,
  Map<String, dynamic>? location,
}) {
  final source = buildGigSearchText(
    title: title,
    description: description,
    location: location,
  );
  final grams = <String>{};
  for (var size = 1; size <= 3; size++) {
    if (source.length < size) break;
    for (var index = 0; index <= source.length - size; index++) {
      grams.add(source.substring(index, index + size));
    }
  }
  final result = grams.toList(growable: false)..sort();
  return result;
}

String gigSearchLookupGram(String term) {
  final normalized = normalizeGigSearchText(term);
  if (normalized.isEmpty) return '';
  return normalized.substring(0, normalized.length < 3 ? normalized.length : 3);
}
